src/Featurizers/AnalyticalRollingWindowFeaturizer.h

```diff
@@ -0,0 +1,307 @@
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+    AnalyticalRollingWindowTransformer(std::uint32_t windowSize, AnalyticalRollingWindowCalculation windowCalculation, std::uint32_t horizon, std::uint32_t minWindowCount = 1) ;
+    AnalyticalRollingWindowTransformer(Archive &ar);
+    ~AnalyticalRollingWindowTransformer(void) override = default;
+
+    FEATURIZER_MOVE_CONSTRUCTOR_ONLY(AnalyticalRollingWindowTransformer);
+
+    bool operator==(AnalyticalRollingWindowTransformer const &other) const;
+    bool operator!=(AnalyticalRollingWindowTransformer const &other) const;
+    void save(Archive & ar) const override;
+
+private:
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Types
+    // |
+    // ----------------------------------------------------------------------
+    using OutputType = std::vector<double>;    
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Private Members
+    // |
+    // ----------------------------------------------------------------------
+    // _windowSizes and _minWindowCount are int64 so that we can have the full positive range of the uint32 it comes
+    // in as but can still support negative numbers.
+    Components::CircularBuffer<InputT>      _buffer;
+    std::int64_t                            _windowSize;
+    std::uint32_t                           _horizon;
+    std::int64_t                            _minWindowCount;
+    AnalyticalRollingWindowCalculation      _windowCalculation;
+
+    // ----------------------------------------------------------------------
+    // |
+    // |  Public Methods
+    // |
+    // ----------------------------------------------------------------------
+
+    // MSVC runs into problems when the declaration and definition are separated
+    void execute_impl(typename BaseType::InputType const &input, typename BaseType::CallbackFunction const &callback) override {
+        OutputType         results(_horizon);
+        
+        _buffer.push(input);
+        size_t bufferSize = _buffer.size();
+
+        for (int i = _horizon; i > 0; i--){
+            // We cast to int32 here because -1 is valid for the algorithm.
+            std::int32_t elementsAvailableToUse(static_cast<std::int32_t>(bufferSize) - i);
+
+            // If we don't have enough elements then output NaN
+            if (elementsAvailableToUse < _minWindowCount) {
+                results[_horizon - i] = std::nan("");
+
+            // If we have strictly less elements then the window size, but more then the minimum size
+            } else if (elementsAvailableToUse < _windowSize) {
+                const auto range = _buffer.range(_windowSize - elementsAvailableToUse);

```

Comment:
>auto [](start = 22, length = 4)

Only use `auto` in these situations:

1) Variable defined within a loop (e.g. a for loop)
2) When declaring an anonymous function
3) In complicated template metaprogramming scenarios.

Reply:
So this auto is in a for loop. I have now replaced it with the actual type anyways. Interesting though is that in the process I ran into an internal compiler error. I was able to work around it, it was just interesting to see.

Reply:
By "Variable defined within a loop", I mean something like:

`for(auto it = container.begin(); it != container.end(); ++it) { ... }`

---
In reply to: [395302156](https://github.com/microsoft/FeaturizersLibrary/pull/173#discussion_r395302156) [](ancestors = 395302156)